### ISSUE TITLE
Final variables in methods

### ISSUE THREAD

**Issue by wei2912** [nus-cs2030/2021-s1#30]
# Description 
In Lab 1 & 2, the concept of using the `final` keyword for a class's attributes was brought up as a way of ensuring an object's immutability. However, the methods in the code provided in this module don't seem to use the `final` keyword for their variables. What's the usual practice adopted in CS2030S for variables in methods?

(See https://stackoverflow.com/questions/8306313/final-variable-in-methods-in-java for a Stack Overflow answer on `final` variables in methods. Other languages such as JavaScript and C also have constructs to create constants.)

## Topic:
Coding practices

## Screenshots (if any): 
NA

**Comment by domlimm** [comment]
Hi, if i'm not wrong the task in Lab 2 uses the `final` keyword. Prof Henry came in earlier today during my lab session and said something along the line that he amended all the lectures/labs/recitations to something else. I forgot what was it. But i think from Lab 2 onwards we should all be seeing the `final` keyword.

Also today, my tutor did mention that we **should always use** the `final` keyword for the attributes in classes when he went through last week's Lab (01).

But yeah it'd be good to know too on CS2030/S usual practice!

**Comment by pchunky** [comment]
When you mentioned "the methods in the code provided in this module", which part of the module's materials were you referring to? Lecture notes? CodeCrunch question prompts?

In any case, CS2030/S enforce immutability. So, use `final` as much as possible. Then again, there is nothing set in stone in programming. Throughout the module, you will see mutable classes, and some un-`final` variables too. So, my advice will be to keep immutability as your number one priority. Have a sense to know when immutability is simply not possible and un-`final` in that case.

**Comment by wei2912** [comment]
@purfectliterature I was referring to both the lecture notes and CodeCrunch question prompts. For example, Recitation 2 has this code snippet:

![image](https://user-images.githubusercontent.com/2070423/92306788-62e49f00-efc4-11ea-96af-142cc1f818ab.png)

The attributes `centre` and `radius` were made `final`, while the `circle` variable in `equals()` is still mutable. I was wondering if this was a code style preference or typo...

Also, thanks @domsterthebot for your answer too!

**Comment by pchunky** [comment]
Are you referring to `equals(Object)` or `equals(Circle)`?

If you are referring to `equals(Object)`, then `Circle circle = (Circle) obj` is just a variable assignment. Changing that to `final Circle circle` will not change the Circle class' immutability, would it? `final` keyword is just a keyword which prevents the change of a variable. That's it. It has nothing to do with immutability. But, we use them in classes' attributes to prevent state mutations. In a function body like this, I don't see the difference/benefit in using `final` for a local variable, so I don't see anything particularly interesting about this code snippet.

If you are referring to `equals(Circle)`, then the `Circle circle` is the method's parameter.

**Comment by wei2912** [comment]
@purfectliterature My apologies for that, I should have made it clearer that I was talking about the immutability of __local variables__, and not of the `Circle` class. As to the example above, I am referring to the local variable `Circle` in the `equals(Object)` method.

On the use of `final` for local variables, I'd agree that there isn't much of a benefit in this case (of the `equals(Object)` and `equals(Circle)` methods) given the small size of the methods. But the same could be said of the class attributes, since the class itself is also rather small and it's very easy to ensure that the internal class methods do not mutate variables (so long as they are `private`). Hence, my question is about what distinguishes these two cases, and why CS2030 enforces **immutable class attributes** but does not enforce **immutable local variables** wherever possible? After all, accidental mutation of variables inside methods is also a common issue - the answers to [this Stack Overflow question](https://softwareengineering.stackexchange.com/questions/115690/why-declare-final-variables-inside-methods) seem to agree that it's a good habit to enforce `final` on variables that should not be mutated.

**Comment by HolmesJJ** [comment]
Hi, @wei2912, actually I don’t really understand your question. The variable with **final** keyword represents a constant, can only be assigned once, and the value will not change after assignment. **final** will be used when you feel that this variable does not need to be changed anymore in the future.
In a word, whether it is necessary to use **final** keyword is based on actual scenarios, there is no regulation.

**Comment by waynetee** [comment]
Hi Wei En,

I don't have strong views one way or another, but let me share my thoughts.

Consider this example:
```
Foo foo = new Foo();
foo.bar();
...
foo.bar();
```
If the class attributes are kept unchanged, then `foo.bar` will return the same thing no matter what happens between them.

This is desirable in some sense, and making the class attributes `final` is a way to enforce this.

Now let's talk about local variables. Unlike class attributes, keeping them constant does not provide any guarantee anything about the external behaviour of the class. I think this would be the difference you are asking for.

Of course, as you pointed out, there are good reasons to declare variables that are supposed to be constant as such.

But coming from my personal point of view as a lazy programmer, if the code is simple, I would rather save some keystrokes and keep the code slightly shorter, and slightly easier to read.

Let me know what you think

**Comment by pchunky** [comment]
Alright. Let me settle this once and for all (hopefully).

>But the same could be said of the class attributes, since the class itself is also rather small and it's very easy to ensure that the internal class methods do not mutate variables (so long as they are private).

Not really. The "size" of a class does not signify anything, really--I mean, how do we quantify the "size" of a class? The aim for using `final` and achieving immutability is not just preventing accidental mutation. Imagine if you are writing a software with a lot packages and classes, (im)mutability becomes extremely important to ensure all dependency and relationship do not break just because one class was mutated when it is not supposed to.


>After all, accidental mutation of variables inside methods is also a common issue

Yeah. It's even worse with cases e.g., `List`s have `.size()` method to **retrieve** the size of the list. In the definition of `List` there is probably an attribute `size` too. The same goes for `length()` for `String`s and `.length` for arrays. If you categorise these cases as __accidental mutation__, then okay.

### Bottom line is 
When we are talking about immutability, we are talking about classes and their attributes. These are **objects** whose states cannot change once being initialised. What you are talking about in the case of local variables is not in the conversation of immutability. It is just your normal syntactical dilemmas.

Ask every JavaScript programmers. They will probably tell you that they use `const`, `let`, (and rarely now) `var`. But, I (and I believe a lot of people) use `const` more often than the other. There are no reasons for us to really change these variables. However, there are some instances where we might use `let` or `var`. **But**, we use `const` not because we learn from CS2030, but because that is a keyword for variable assignment. So, might as well.

In the case for `Circle circle = (Circle) obj`, `final` is optional. This is different from the principles of **immutability**, in which `final` is mandatory (as much as possible). If you want to use `final` in this casting, sure. If not, sure. After all, can the value of this variable change in the function itself? No, right? So, why bother?

What is interesting from your question is the phrase **"CS2030 enforcing"**. I use `private final` and choose my access modifiers not because my module enforces me, but because I am following the principles of good software design, which will definitely help me as I extend this software. In this `Circle` example, sure, one can argue that it's a simple program, and no chance of mutability. Sure. But this is a module, and **we train students**. So, we start from the simplest, and hope that students will build the conscience to write well. As you go through the module, see more labs, and do the projects(!), these principles become more important than ever.

**TL;DR: Want to use `final` for local variable can, don't want to use also can. Immutability here we talk about is in terms of classes, not variable assignments.**

Cheers and hope it helps. Remember to wash your hands at least 20 seconds.

**Comment by wei2912** [comment]
@purfectliterature Thanks for the clarification, that was a very helpful writeup!

EDIT: Also, thanks to @HolmesJJ and @waynetee for your replies too. Personally I would still prefer to enforce `final` variables if there's no need for me to modify them, but I can see the significance of ensuring class immutability and how it's different from the issue of mutable local variables.

**Comment by pchunky** [comment]
You're most welcome. It's good that you are conscious about using `final`. Just know when to and when not to ><

~~But probably sooner or later you will find it to be a hassle.~~
